### ISSUE TITLE
Shouldn't jooq-codegen be exposed through the api configuration?

### ISSUE THREAD

**Issue by Double-O-Seven** [etiennestuder/gradle-jooq-plugin#143]
Hello,

I'm using this plugin to create a pre-compiled Kotlin DSL build script.

`buildSrc` contains `implementation(group = "nu.studer", name = "gradle-jooq-plugin", version = "5.1.0")` as dependency. The script itself uses `id("nu.studer.jooq")` to apply the plugin. So far so good, extension is recognized. However, importing the following for example doesn't work:
```
import org.jooq.meta.jaxb.ForcedType
import org.jooq.meta.jaxb.Logging
import org.jooq.meta.jaxb.Property
```

I need to manually add `implementation(group = "org.jooq", name = "jooq-codegen", version = "3.13.4")` as dependency and then the imports are resolved.

If the plugin would declare `jooq-codegen` as dependency in the `api` configuration, the problem would be solved I assume?

I'm on Gradle 6.6.1 by the way.

Thanks in advance!

**Comment by etiennestuder** [comment]
I don't understand what you are trying to achieve. Why do you pull in the plugin as a normal dependency? Why do you apply the plugin in your script and not in your pre-compiled plugin? Please point me to a repo with a full example.

See this example where the imports are available:
https://github.com/etiennestuder/gradle-jooq-plugin/tree/master/example/use_kotlin_dsl

**Comment by Double-O-Seven** [comment]
Ah, the plugin was of course applied in the precompiled script.

If I pack the example script 1:1 into a precompiled script, I get the following error:
```Invalid plugin request [id: 'nu.studer.jooq', version: '5.1.0']. Plugin requests from precompiled scripts must not include a version number. Please remove the version from the offending request and make sure the module containing the requested plugin 'nu.studer.jooq' is an implementation dependency of project ':buildSrc'.
```

Now, if I remove the version declaration from the plugin block, I get another error:
```
Caused by: org.gradle.api.plugins.UnknownPluginException: Plugin [id: 'nu.studer.jooq'] was not found in any of the following sources:
```

This error itself can be fixed by declaring the following dependency in the `build.gradle.kts` file of `buildSrc`:
```implementation("nu.studer:gradle-jooq-plugin:5.1.0")```

Now, the plugin is recognized, however, the imported  JOOQ classes are not recognized.

The only way to fix it is by declaring yet another dependency in the `build.gradle.kts` file of `buildSrc`:
```implementation("org.jooq:jooq-codegen:3.13.4")```

So to summarize:
Precompiled script cannot access JOOQ classes, possibly because `jooq-codegen` is an `implementation` dependency of your plugin.

**Comment by etiennestuder** [comment]
> Precompiled script cannot access JOOQ classes, possibly because `jooq-codegen` is an `implementation` dependency of your plugin.

That interpetation is definitely not true. implementation/api is specific to compile classpath visibility by dependent projects, and not related to visibility within build scripts.

**Comment by etiennestuder** [comment]
See this working [example](https://github.com/etiennestuder/gradle-jooq-plugin/tree/78676876239a578eec20109d12ff1aea91941642/example/extract_precompiled_script_plugin).

How to write precompiled script plugins is described [here](https://docs.gradle.org/current/userguide/custom_plugins.html#sec:precompiled_plugins).

> Precompiled script cannot access JOOQ classes, possibly because jooq-codegen is an implementation dependency of your plugin.

Could be. I did not investigate further given the simple workaround.

**Comment by etiennestuder** [comment]
I've brought this to the attention of our Core team. They are investigating whether this is a bug in Gradle or not.

**Comment by etiennestuder** [comment]
> So to summarize:
> Precompiled script cannot access JOOQ classes, possibly because `jooq-codegen` is an `implementation` dependency of your plugin.

Turns out your assumption was spot on. I've [released](https://github.com/etiennestuder/gradle-jooq-plugin/releases/tag/v5.1.1) 5.1.1 which fixes the issue, meaning you do not have to declare the dependency on the `org.jooq:jooq-codegen` library in the build of the precompiled script plugin anymore.

**Comment by etiennestuder** [comment]
Thanks for having reported and investigated this.

**Comment by Double-O-Seven** [comment]
Great, thanks!
